src/mlpack/methods/ann/layer/spatial_dropout_impl.hpp

```diff
@@ -0,0 +1,127 @@
+/**
+ * @file methods/ann/layer/spatial_dropout_impl.hpp
+ * @author Anjishnu Mukherjee
+ *
+ * Implementation of the SpatialDropout class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_SPATIAL_DROPOUT_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_SPATIAL_DROPOUT_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "spatial_dropout.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+SpatialDropout<InputDataType, OutputDataType>::SpatialDropout() :
+    size(0),
+    ratio(0.5),
+    scale(1.0 / (1.0 - ratio)),
+    reset(false),
+    batchSize(0),
+    inputSize(0),
+    deterministic(false)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+SpatialDropout<InputDataType, OutputDataType>::SpatialDropout(
+    size_t size,
+    double ratio) :
+    size(size),
+    ratio(ratio),
+    scale(1.0 / (1.0 - ratio)),
+    reset(false),
+    batchSize(0),
+    inputSize(0),
+    deterministic(false)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void SpatialDropout<InputDataType, OutputDataType>::Forward(
+  const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  if (!reset)
+  {
+    batchSize = input.n_cols;
+    inputSize = input.n_rows / size;
+    reset = true;
+  }
+
+  if (deterministic)
+    output = input;
+
+  else
+  {
+    output.zeros(arma::size(input));
+    arma::cube inputTemp(const_cast<arma::mat&>(input).memptr(), inputSize,
+        size, batchSize, false, false);
+    arma::cube outputTemp(const_cast<arma::mat&>(output).memptr(), inputSize,
+        size, batchSize, false, false);
+    arma::mat probabilities(1, size);
+    arma::mat maskRow(1, size);
+    probabilities.fill(ratio);
+    ann::BernoulliDistribution<> bernoulli_dist(probabilities, false);
+    maskRow = bernoulli_dist.Sample();
+    mask = arma::repmat(maskRow, inputSize, 1);
+
+    for (size_t n = 0; n < batchSize; n++)
+    {
+      arma::mat& inputImage = inputTemp.slice(n);
+      arma::mat& outputImage = outputTemp.slice(n);
+      outputImage = inputImage % mask * scale;
```

Comment:
Maybe I missed something, but if this is the only place where we use `inputImage` and `outputImage` I would directly use `inputTemp.slice(n)` and `outputTemp.slice(n)`.